### PR TITLE
Aikepotze/pytorch base model

### PR DESCRIPTION
**Pull request by aikepotze**
Implemented base NN model class and example LSTM implementation. Fit, predict_batch and predict functionalities of the example LSTM are covered by new test.

### CODE REVIEW

**Review by krsnapaudel — CHANGES_REQUESTED** [review]
`fit` and `predict_batch` have new signatures. Does this need update to the base class?
See here a discussion:
https://stackoverflow.com/questions/41235517/method-overriding-can-have-different-number-of-parameters-in-subclass-and-parent

File: models/nn_models.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        # Set default values for loss_fn, optim_fn, optim_kwargs, scheduler_fn, scheduler_kwargs
+        if loss_fn is None: loss_fn = torch.nn.functional.mse_loss
+        if loss_kwargs is None: loss_kwargs = {"reduction": "mean"}
+        if optim_fn is None: optim_fn = torch.optim.Adam
+        if optim_kwargs is None: optim_kwargs = {}
+
+        assert num_epochs > 0
+    
+        # Send model to device
+        self.to(device) #TODO this doesn't make sense
```
Comment (michielkallenberg) [inline models/nn_models.py:71]: :-)

File: models/nn_models.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                pbar.set_description(
+                    f"Epoch {epoch+1}/{num_epochs} | Loss: {loss.item():.4f}"
+                )
+
+
+            if val_loader is not None and (epoch+1) % val_every_n_epochs == 0:
+                self.eval()
+
+                # Validation loop
+                with torch.no_grad():
```
Comment (michielkallenberg) [inline models/nn_models.py:143]: May want to set the model to "eval" here as well (this will be relevant in case you'll add dropout or batch norm later), and then after evaluation has been done set it back to "train"

File: tests/models/test_model.py [hunk 3]
```diff
@@ -96,3 +99,51 @@ def test_sklearn_model():
     model.fit(train_dataset, **fit_params)
     test_preds, _ = model.predict(test_dataset)
     assert test_preds.shape[0] == len(test_dataset)
+
+def test_nn_model():
+
+    train_dataset = Dataset.load("test_maize_us")
```
Comment (michielkallenberg) [inline tests/models/test_model.py:105]: sure about "test_maize_us"; not "train_maize_us" ?

**Review by michielkallenberg — CHANGES_REQUESTED** [review]

**Comment by aikepotze** [comment]
> `fit` and `predict_batch` have new signatures. Does this need update to the base class? See here a couple of discussions: https://stackoverflow.com/questions/154577/polymorphism-vs-overriding-vs-overloading https://stackoverflow.com/questions/41235517/method-overriding-can-have-different-number-of-parameters-in-subclass-and-parent
> 
> What we want to is overriding the `BaseModel`, not overloading.
> 
> @ronvree and I discussed and the overriding of `fit` is probably fine, but `predict_batch` is not. We can add `**kwargs` to `predict*` methods. Create a data loader in `predict`. Use default batch size (e.g. 16 or 32) instantiated in the constructor. This can be updated if `fit` gets a batch size argument.

I've changed the signature of predict_batch batch to the original version. Now it uses the batch_size and device stored in .fit by default. Users can change device manually anyways outside of predict_batch method.

**Review by krsnapaudel — APPROVED** [review]
For now the way we select static and time series data is fine. We need to add constants to `config.py` to list keys that identify statics vs time series variables. Remove `main` from test file and merge.

**Comment by aikepotze** [comment]
Removed main from test file. Addressed all comments by @michielkallenberg, please take a look again so I can merge the pull request.

File: models/nn_models.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    def forward(self, x):
+        max_n_dim = max([len(v.shape) for k, v in x.items() if isinstance(v, torch.Tensor)])
+        x_ts = torch.cat([v.unsqueeze(2) for k, v in x.items() if isinstance(v, torch.Tensor) and len(v.shape) == max_n_dim], dim=2)
+        x_static = torch.cat([v.unsqueeze(1) for k, v in x.items() if isinstance(v, torch.Tensor) and len(v.shape) < max_n_dim], dim=1)
+
+        x_ts, _ = self._lstm(x_ts)
+        x = torch.cat([x_ts[:, -1, :], x_static], dim=1)
+        x = self._fc(x)
+        return x
+
```
Comment (michielkallenberg) [inline models/nn_models.py:226]: newline party :-)

File: tests/models/test_model.py (hunk 3 above)
Comment (michielkallenberg) [reply tests/models/test_model.py:190]: @aikepotze can you have one more look at this concern?

**Review by michielkallenberg — APPROVED** [review]
